src/js_subscriptions.hpp

```diff
@@ -0,0 +1,815 @@
+
+    MethodMap<T> const methods = {
+        {"snapshot", wrap<snapshot>},
+        {"findByName", wrap<find_by_name>},
+        {"find", wrap<find>},
+        {"update", wrap<update>},
+        {"_waitForSynchronization", wrap<wait_for_synchronization>},
+    };
+};
+
+template <typename T>
+typename T::Object SubscriptionsClass<T>::create_instance(ContextType ctx,
+                                                          realm::sync::SubscriptionSet subscriptionSet)
+{
+    return create_object<T, SubscriptionsClass<T>>(ctx, new Subscriptions<T>(std::move(subscriptionSet)));
+}
+
+/**
+ * @brief Get whether the subscriptions collection is empty or not
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param return_value \ref ReturnValue wrapping a boolean containing the empty state
+ */
+template <typename T>
+void SubscriptionsClass<T>::get_empty(ContextType ctx, ObjectType this_object, ReturnValue& return_value)
+{
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+    return_value.set(subs->size() == 0);
+}
+
+/**
+ * @brief Get the error string for the subscription set, if any
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param return_value \ref ReturnValue wrapping the error string if any, or null if not
+ */
+template <typename T>
+void SubscriptionsClass<T>::get_error(ContextType ctx, ObjectType this_object, ReturnValue& return_value)
+{
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+    std::string error = subs->error_str();
+
+    if (error == "") {
+        return_value.set_null();
+    }
+    else {
+        return_value.set(error);
+    }
+}
+
+/**
+ * @brief Get the current state of the subscription set
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param return_value \ref ReturnValue wrapping a string representing the current state
+ * @exception std::runtime_error if an unknown state is encountered
+ */
+template <typename T>
+void SubscriptionsClass<T>::get_state(ContextType ctx, ObjectType this_object, ReturnValue& return_value)
+{
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+    switch (subs->state()) {
+        case sync::SubscriptionSet::State::Uncommitted:
+        case sync::SubscriptionSet::State::Pending:
+        case sync::SubscriptionSet::State::Bootstrapping:
+            return_value.set("pending");
+            break;
+        case sync::SubscriptionSet::State::Complete:
+            return_value.set("complete");
+            break;
+        case sync::SubscriptionSet::State::Error:
+            return_value.set("error");
+            break;
+        case sync::SubscriptionSet::State::Superceded:
+            return_value.set("superceded");
+            break;
+        default:
+            throw std::runtime_error("Unknown state in get_state()");
```

Comment:
I think `REALM_UNREACHABLE()` is better.

Reply:
👍